// angular2-ladda/esm2015/lib/ladda.module.js
import * as ɵngcc0 from '@angular/core';
var LaddaModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { LaddaDirective } from './ladda.directive';
import { LaddaConfig } from './ladda-config';
let LaddaModule = LaddaModule_1 = class LaddaModule {
    static forRoot(config) {
        return {
            ngModule: LaddaModule_1,
            providers: [
                { provide: LaddaConfig, useValue: config }
            ]
        };
    }
};
LaddaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LaddaModule });
LaddaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LaddaModule_Factory(t) { return new (t || LaddaModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LaddaModule, { declarations: function () { return [LaddaDirective]; }, exports: function () { return [LaddaDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LaddaModule, [{
        type: NgModule,
        args: [{
                declarations: [LaddaDirective],
                exports: [LaddaDirective]
            }]
    }], null, null); })();
export { LaddaModule };
